File: 043_LRUCache.cpp
// Topic: LRU Cache
// Description: Design a cache supporting get and put in O(1) time with LRU eviction.
// Approach: Use unordered_map for key->iterator and doubly-linked list (std::list) to maintain
// recency.

#include <bits/stdc++.h>
using namespace std;

class LRUCache
{
    int cap;
    list<pair<int, int>> dq;  // {key, value}, front = most recent
    unordered_map<int, list<pair<int, int>>::iterator> mp;

   public:
    LRUCache(int capacity) : cap(capacity) { mp.reserve(capacity * 2); }
    int get(int key)
    {
        auto it = mp.find(key);
        if (it == mp.end())
            return -1;
        auto nodeIt = it->second;
        int val = nodeIt->second;
        dq.erase(nodeIt);
        dq.push_front({key, val});
        mp[key] = dq.begin();
        return val;
    }
    void put(int key, int value)
    {
        auto it = mp.find(key);
        if (it != mp.end())
        {
            dq.erase(it->second);
            mp.erase(it);
        }
        dq.push_front({key, value});
        mp[key] = dq.begin();
        if ((int)mp.size() > cap)
        {
            auto last = dq.back();
            mp.erase(last.first);
            dq.pop_back();
        }
    }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    LRUCache cache(2);
    cache.put(1, 1);
    cache.put(2, 2);
    cout << "get(1) -> " << cache.get(1) << "\n";  // returns 1
    cache.put(3, 3);
    cout << "get(2) -> " << cache.get(2) << "\n";  // returns -1
    cache.put(4, 4);
    cout << "get(1) -> " << cache.get(1) << "\n";  // returns -1
    cout << "get(3) -> " << cache.get(3) << "\n";  // returns 3
    cout << "get(4) -> " << cache.get(4) << "\n";  // returns 4
}

/*
Time Complexity: O(1) per operation
Space Complexity: O(capacity)
*/

File: 043_LRUCache.py
# Topic: LRU Cache
# Description: Design a cache supporting get and put in O(1) time with LRU eviction.
# Approach: Use dict for key->node and doubly-linked list (OrderedDict) to maintain recency.

from collections import OrderedDict


class LRUCache:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache = OrderedDict()

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        self.cache.move_to_end(key)
        return self.cache[key]

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache.move_to_end(key)
        self.cache[key] = value
        if len(self.cache) > self.cap:
            self.cache.popitem(last=False)


if __name__ == "__main__":
    cache = LRUCache(2)
    cache.put(1, 1)
    cache.put(2, 2)
    print(f"get(1) -> {cache.get(1)}")   # 1
    cache.put(3, 3)
    print(f"get(2) -> {cache.get(2)}")   # -1
    cache.put(4, 4)
    print(f"get(1) -> {cache.get(1)}")   # -1
    print(f"get(3) -> {cache.get(3)}")   # 3
    print(f"get(4) -> {cache.get(4)}")   # 4

"""
Time Complexity: O(1) per operation
Space Complexity: O(capacity)
"""
